### ISSUE TITLE
Line ends (arrow heads and others) missing or misplaced in qualified associations

### ISSUE THREAD

**Issue by travkin79** [plantuml/plantuml#1491]
**Problem description**
We've found a few cases where arrow heads (and other line ends) are not correctly drawn, i.e. are misplaced or not visible at all. That seems to affect only qualified associations (which are really tricky) and a few rare cases in component diagrams.

**Steps to reproduce**
Create a file with the following diagram examples and render them with PlantUML version 1.2023.10.

**Example 1**

*Diagram source code*
```plantuml
@startuml
    class HashMap<K,V>

    HashMap [key: K] -u-> V1
    HashMap [key: K] -r-> V2
    HashMap [key: K] -d-> V3
    HashMap [key: K] -l-> V4
@enduml
```

*Rendered diagram*
![image](https://github.com/plantuml/plantuml/assets/10887297/0b67c772-b52b-4801-8690-54f7e912285d)

*Expected diagram*
![image](https://github.com/plantuml/plantuml/assets/10887297/994d5217-a73b-47fa-9953-2ac0fc1d0b21)

**Example 2**

*Diagram source code*
```plantuml
@startuml
    class HashMap<K,V>
    HashMap [a1] <|-u-> [e] V1
    HashMap [b2] *.r.> [f] V2
    HashMap [c3] o.d.> [g] V3
    HashMap [d4] +-l-> [h] V4

    MoreComplex [x: String] ..> X
    MoreComplex [y: int] --> Y
    MoreComplex [z: boolean] ..> Z
@enduml
```

*Rendered diagram*
![image](https://github.com/plantuml/plantuml/assets/10887297/f1977ee3-f19b-4a60-839e-7086ac0165a9)
Here, the two horizontal line's ends on the `HashMap` side (filled diamond and circle with an X) are completely misplaced. They are drawn on the classes `V1` and `V3`. The other line ends are not drawn as starting / ending on the qualifier labels, but on the source and target classes. The qualifier labels are not sticked to the involved classes, but seem to jump away from them to make space for the arrow ends.

*Expected diagram*
In case of these associations with qualifiers on both ends, the lines with their line endings (arrow heads and others) should be drawn between the two qualifier labels. The qualifier labels are to be placed on the side of the corresponding classes like in example 1.

**Example 3**

*Diagram source code*
```plantuml
@startuml
 
skinparam fixCircleLabelOverlapping true
 
() "Product Dubbo" as product_dubbo
() "Member Dubbo" as member_dubbo
() "Guide Dubbo" as guide_dubbo
() "Search Dubbo" as search_dubbo
 
[ME]
[P] - product_dubbo
member_dubbo - [M]
[G] - guide_dubbo
search_dubbo - [S]
 
guide_dubbo <. [ME]
[ME] ..> member_dubbo
[ME] ..> product_dubbo
[ME] .> search_dubbo

footer Rendered with PlantUML version %version()
 
@enduml
```

*Rendered diagram*
![image](https://github.com/plantuml/plantuml/assets/10887297/b4992751-eff4-46bc-96bd-322a65350a56)

*Expected diagram*
![image](https://github.com/plantuml/plantuml/assets/10887297/695fa62c-227c-4e35-a0c8-3eccdcf86c12)

**Execution environment**
OS: macOS Ventura 13.4
JRE: OpenJDK Runtime Environment Temurin-11.0.15+10 (build 11.0.15+10)
GraphViz: dot - graphviz version 8.0.5 (20230430.1635)

**Comment by arnaudroques** [comment]
It should [work better now](https://www.plantuml.com/plantuml/uml/PS-_geCm5CVnFKznvdkRW7yc4H6w58fJGTfAXgXf5M8YIGHR-l0zjdGGC_u-oHUIEo-idmP504-ZX7Dm4Auhn9WUTsMspBy1AX5mI5zqeXbKaaCPhBaEEVmnor1lgE5QcuY3OUsYTzHehMtCuP-gfToXnkJ3XR5oRuPHoHcgEO6Jjxt6rmmhv-sDHmAzzXmePYvRVYPG6wEaqD_2bH2IIztYzzy0).

However, we slightly change the way we send data to GraphViz so there might be some regression (even when there are no qualified associations).

Any comment is welcome then! :-)

**Comment by travkin79** [comment]
Hi Arnaud,
Thanks for your work.

I opened the link from your comment and I wondered why the layout still looked odd. Then, I realized that for some reason the online server shows an improved layout for only about 1 second and then changes the displayed image to something else. Can you reproduce that?

The image I see for only about 1 second:
![image](https://github.com/plantuml/plantuml/assets/10887297/5bc7e6e6-a0f2-49b1-9741-673792010671)
That looks great of course.

Then, without doing anything, I suddenly see this image:
![image](https://github.com/plantuml/plantuml/assets/10887297/7d54272e-27ee-45b7-b9b4-23a67cea9d9d)

I'll try to test your changes on my machine with some other diagrams, too.

**Comment by travkin79** [comment]
Hi Arnaud,
I switched to using the current snapshot version 1.2023.11beta2 in my tools and went through quite a few various diagrams (class, component, object diagrams, ...) rendering the images. I did not found any problems with your change. Really great work! Thank you, @arnaudroques.

There was only one minor issue with using qualified associations and generics / templates in class diagrams. In very rare cases the qualifier may by placed on top of the generics label like in the following [example](https://www.plantuml.com/plantuml/uml/JP0zIyOm58Nt_ef3GL1e0wvIGi55iGLHTAaEeRsrWJGf-QYBFzuqRmDlXX1EdkSUkAtpmlgmgWBnIEt9pc8az69hNgfFdk9H2UVm9Dnon9tHFzLZSDwiP7cHY9YZ-MCiGobyNyo6OJp9v-a1AghVgEjGSvJvuqlVbyZp3Z4FmI2ZqrqSUoYxK86kuTp264UkJTTiZ2UBDz8JMPhmA_s2LoMq_-Wxx6ITD1etv-FshcWZ6TVm3m00). But I can live with that and would suggest to close this issue.

<img width="264" alt="image" src="https://github.com/plantuml/plantuml/assets/10887297/a37d63f8-455c-4821-9d59-0478372521e1">
